File: autocontext/src/autocontext/storage/context_selection_store.py
from __future__ import annotations

import re
from pathlib import Path
from typing import TYPE_CHECKING, Any

from autocontext.knowledge.context_selection import SCHEMA_VERSION, ContextSelectionDecision
from autocontext.storage.run_paths import resolve_run_root
from autocontext.util.json_io import read_json

if TYPE_CHECKING:
    from autocontext.storage.artifacts import ArtifactStore

_SAFE_STAGE_RE = re.compile(r"[A-Za-z0-9_.-]+")
_DECISION_FILE_RE = re.compile(r"gen_(?P<generation>[0-9]+)_(?P<stage>[A-Za-z0-9_.-]+)\.json\Z")


def context_selection_decision_path(
    runs_root: Path,
    decision: ContextSelectionDecision,
) -> Path:
    run_root = resolve_run_root(runs_root, decision.run_id)
    if decision.generation < 0:
        raise ValueError(f"generation must be non-negative: {decision.generation!r}")
    if not _SAFE_STAGE_RE.fullmatch(decision.stage):
        raise ValueError(f"stage must be a single safe path segment: {decision.stage!r}")
    return run_root / "context_selection" / f"gen_{decision.generation}_{decision.stage}.json"


def persist_context_selection_decision(
    artifacts: ArtifactStore,
    decision: ContextSelectionDecision,
) -> Path:
    path = context_selection_decision_path(artifacts.runs_root, decision)
    artifacts.write_json(path, decision.to_dict())
    return path


def load_context_selection_decisions(
    runs_root: Path,
    run_id: str,
) -> list[ContextSelectionDecision]:
    context_dir = resolve_run_root(runs_root, run_id) / "context_selection"
    if not context_dir.exists():
        return []
    decisions: list[ContextSelectionDecision] = []
    for path in sorted(context_dir.glob("gen_*_*.json")):
        match = _DECISION_FILE_RE.fullmatch(path.name)
        if match is None:
            continue
        data = read_json(path)
        decision = _decision_from_payload(
            data,
            run_id=run_id,
            generation=int(match.group("generation")),
            stage=match.group("stage"),
        )
        if decision is not None:
            decisions.append(decision)
    return sorted(decisions, key=lambda decision: (decision.generation, decision.stage))


def _decision_from_payload(
    data: Any,
    *,
    run_id: str,
    generation: int,
    stage: str,
) -> ContextSelectionDecision | None:
    if not isinstance(data, dict):
        return None
    if data.get("schema_version") != SCHEMA_VERSION:
        return None
    if data.get("run_id") != run_id:
        return None
    if type(data.get("generation")) is not int or data.get("generation") != generation:
        return None
    if data.get("stage") != stage or not _SAFE_STAGE_RE.fullmatch(stage):
        return None
    if not isinstance(data.get("scenario_name"), str):
        return None
    if not isinstance(data.get("candidates"), list):
        return None
    if not _has_decision_metrics(data.get("metrics")):
        return None
    return ContextSelectionDecision.from_dict(data)


def _has_decision_metrics(value: Any) -> bool:
    if not isinstance(value, dict):
        return False
    required_keys = {
        "candidate_count",
        "selected_count",
        "candidate_token_estimate",
        "selected_token_estimate",
    }
    return required_keys.issubset(value)

File: ts/src/knowledge/context-selection-store.ts
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { isAbsolute, join, relative, resolve } from "node:path";

import type { ContextSelectionDecisionInput } from "./context-selection-report.js";

const SCHEMA_VERSION = 1;
const SAFE_STAGE_RE = /^[A-Za-z0-9_.-]+$/;
const DECISION_FILE_RE = /^gen_(?<generation>[0-9]+)_(?<stage>[A-Za-z0-9_.-]+)\.json$/;

export function loadContextSelectionDecisions(
  runsRoot: string,
  runId: string,
): ContextSelectionDecisionInput[] {
  const cleanRunId = runId.trim();
  const contextDir = join(resolveRunRoot(runsRoot, cleanRunId), "context_selection");
  if (!existsSync(contextDir)) {
    return [];
  }
  const decisions: ContextSelectionDecisionInput[] = [];
  for (const fileName of readdirSync(contextDir).sort()) {
    const match = DECISION_FILE_RE.exec(fileName);
    if (!match?.groups) continue;
    const data = JSON.parse(readFileSync(join(contextDir, fileName), "utf-8"));
    const decision = decisionFromPayload(data, {
      runId: cleanRunId,
      generation: Number.parseInt(match.groups.generation!, 10),
      stage: match.groups.stage!,
    });
    if (decision) {
      decisions.push(decision);
    }
  }
  return decisions.sort((a, b) =>
    coerceNumber(a.generation) - coerceNumber(b.generation) ||
    String(a.stage ?? "").localeCompare(String(b.stage ?? "")));
}

function resolveRunRoot(runsRoot: string, runId: string): string {
  const normalized = runId.trim();
  if (!normalized) {
    throw new Error("run_id is required");
  }
  const root = resolve(runsRoot);
  const candidate = resolve(root, normalized);
  if (candidate === root) {
    throw new Error(`run_id must name a run subdirectory: ${runId}`);
  }
  const relativePath = relative(root, candidate);
  if (relativePath === "" || relativePath.startsWith("..") || isAbsolute(relativePath)) {
    throw new Error(`run_id escapes runs root: ${runId}`);
  }
  return candidate;
}

function decisionFromPayload(
  data: unknown,
  expected: { runId: string; generation: number; stage: string },
): ContextSelectionDecisionInput | null {
  if (!isRecord(data)) return null;
  if (data.schema_version !== SCHEMA_VERSION) return null;
  if (data.run_id !== expected.runId) return null;
  if (!Number.isInteger(data.generation) || data.generation !== expected.generation) return null;
  if (data.stage !== expected.stage || !SAFE_STAGE_RE.test(expected.stage)) return null;
  if (typeof data.scenario_name !== "string") return null;
  if (!Array.isArray(data.candidates)) return null;
  if (!hasDecisionMetrics(data.metrics)) return null;
  return data as ContextSelectionDecisionInput;
}

function hasDecisionMetrics(value: unknown): boolean {
  if (!isRecord(value)) return false;
  return [
    "candidate_count",
    "selected_count",
    "candidate_token_estimate",
    "selected_token_estimate",
  ].every((key) => Object.prototype.hasOwnProperty.call(value, key));
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function coerceNumber(value: unknown): number {
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}
